packages/@aws-cdk/aws-iam/lib/private/merge-statements.ts

```diff
@@ -0,0 +1,176 @@
+const LENGTH_CACHE_SYM = Symbol();
+
+type IamValue = unknown | unknown[];
+
+/**
+ * Only the parts of the policy schema we're interested in
+ */
+interface StatementSchema {
+  readonly Sid?: string;
+  readonly Effect?: string;
+  readonly Principal?: IamValue | Record<string, IamValue>;
+  readonly Resource?: IamValue;
+  readonly Action?: IamValue;
+}
+
+/**
+ * Merge as many statements as possible to shrink the total policy doc, modifying the input array in place
+ */
+export function mergeStatements(statements: StatementSchema[]): StatementSchema[] {
+  let merges = findMerges(statements);
+  while (merges.length > 0) {
+    merges.sort((a, b) => a.sizeDelta - b.sizeDelta); // Most negative number first
+    if (merges[0].sizeDelta >= 0) { break; } // Nothing more to be gained
+
+    // Apply all merges, order biggest to smallest gain
+    while (merges.length > 0 && merges[0].sizeDelta < 0) { // Only while we're not adding size
+      const merge = merges.shift()!;
+      statements[merge.i0] = merge.combined;
+      statements.splice(merge.i1, 1);
+
+      // This invalidates all merges involving i0 or i1, and adjusts all indices > i1
+      const invalidIndices = [merge.i0, merge.i1];
+      let j = 0;
+      while (j < merges.length) {
+        if (invalidIndices.includes(merges[j].i0) || invalidIndices.includes(merges[j].i1)) {
+          merges.splice(j, 1);
+        } else {
+          if (merges[j].i0 > merge.i1) { merges[j].i0 -= 1; }
+          if (merges[j].i1 > merge.i1) { merges[j].i1 -= 1; }
+          j++;
+        }
+      }
+    }
+
+    merges = findMerges(statements);
+  }
+  return statements;
+}
+
+function findMerges(statements: StatementSchema[]): StatementMerge[] {
+  const ret = new Array<StatementMerge>();
+  for (let i0 = 0; i0 < statements.length; i0++) {
+    for (let i1 = i0 + 1; i1 < statements.length; i1++) {
+      tryMerge(statements, i0, i1, ret);
+    }
+  }
+  return ret;
+}
+
+function tryMerge(statements: StatementSchema[], i0: number, i1: number, into: StatementMerge[]) {
+  const a = statements[i0];
+  const b = statements[i1];
+  if (a.Effect !== 'Allow' || b.Effect !== 'Allow') { return; }
+  if (a.Sid || b.Sid) { return; }
+
+  const beforeLen = jsonLength(a) + jsonLength(b);
+
+  tryMerging('Resource');
+  tryMerging('Action');
+  tryMerging('Principal');
```

Comment:
IAM noob question: can users include both Action and NotAction keys, both Principal and NotPrincipal keys, etc? If not, does that caught somewhere else in the IAM policy synthesis / would it be worth adding a check here?

Reply:
It cannot, and that is checked elsewhere already.